swaybar/tray/dbusmenu.c

```diff
@@ -0,0 +1,1367 @@
+	} else {
+		sway_log(SWAY_ERROR, "%s%s failed to get IconThemePath: %s",
+			menu->sni->service, menu->sni->menu, strerror(-ret));
+		free(slot);
+	}
+
+	swaybar_dbusmenu_setup_signals(menu);
+	swaybar_dbusmenu_get_layout_root(menu);
+}
+
+void swaybar_dbusmenu_open(struct swaybar_sni *sni,
+		struct swaybar_output *output, struct swaybar_seat *seat, uint32_t serial, 
+		int x, int y) {
+	struct swaybar_dbusmenu *dbusmenu = sni->tray->menu;
+	if (!dbusmenu) {
+		dbusmenu = calloc(1, sizeof(struct swaybar_dbusmenu));
+		if (!dbusmenu) {
+			sway_log(SWAY_DEBUG, "Could not allocate dbusmenu");
+			return;
+		}
+		sni->tray->menu = dbusmenu;
+	}
+
+	dbusmenu->sni = sni;
+	dbusmenu->output = output;
+	dbusmenu->seat = seat;
+	dbusmenu->serial = serial;
+	dbusmenu->x = seat->pointer.x;
+	dbusmenu->y = seat->pointer.y;
+	dbusmenu->bar = output->bar;
+
+	swaybar_dbusmenu_setup(dbusmenu);
+}
+
+static void close_child_menus_except(struct swaybar_dbusmenu_menu *menu,
+		int id) {
+	if (!menu || !menu->items) {
+		return;
+	}
+	// close all child menus of menu, except the child menu with the given id
+	for (int i = 0; i < menu->items->length; ++i) {
+		struct swaybar_dbusmenu_menu_item *item = menu->items->items[i];
+		if (item->id == id) {
+			continue;
+		}
+		struct swaybar_dbusmenu_menu *menu = item->submenu;
+		if (menu && menu->item_id != 0) {
+			close_menus(menu);
+		}
+	}
+}
+
+static void
+pointer_motion_process_item(struct swaybar_dbusmenu_menu *focused_menu,
+		struct swaybar_dbusmenu_menu_item *item, struct swaybar_seat *seat) {
+	int scale = focused_menu->dbusmenu->output->scale;
+	double x = seat->pointer.x * scale;
+	double y = seat->pointer.y * scale;
+	if (is_in_hotspot(&item->hotspot, x, y) && item->enabled &&
+			!item->is_separator) {
+		struct swaybar_tray *tray = focused_menu->dbusmenu->sni->tray;
+		struct swaybar_sni *sni = tray->menu->sni;
+		if (focused_menu->last_hovered_item != item) {
+			sd_bus_call_method_async(tray->bus, NULL, sni->service, sni->menu, 
+				menu_interface, "Event", NULL, NULL, "isvu", item->id, "hovered",
+				"y", 0, time(NULL));
+
+			sway_log(SWAY_DEBUG, "%s%s hovered id %d", sni->service, sni->menu,
+				item->id);
+
+			// open child menu if current item has a child menu and close other 
+			// potential open child menus. Only one child menu can be open at a time
+			close_child_menus_except(focused_menu, item->id);
+			open_menu_id(focused_menu->dbusmenu, item->id);
+			focused_menu->last_hovered_item = item;
+
+			// a different item needs to be highlighted
+			swaybar_dbusmenu_draw_menu(focused_menu, focused_menu->item_id, false);
+		}
+
+	}
+}
+
+bool dbusmenu_pointer_motion(struct swaybar_seat *seat, 
+		struct wl_pointer *wl_pointer, uint32_t time_, wl_fixed_t surface_x,
+		wl_fixed_t surface_y) {
+	struct swaybar_tray *tray = seat->bar->tray;
+	struct swaybar_dbusmenu_menu *focused_menu = tray->menu_pointer_focus;
+	if (!(tray && tray->menu && focused_menu)) {
```

Comment:
I've started testing this branch out, and I just got a segmentation fault here due to `tray` being NULL, so this `!tray` check may need to occur sooner.

However, the context was that I was running a bar without a tray. I'm not entirely sure why this callback even runs in that case, it would be good to check whether it makes sense to try to disable it.

Reply:
Could you test if my patch [below](https://github.com/swaywm/sway/pull/6249#issuecomment-1543543643) alleviates the issue? The null check should probably still be fixed, otherwise it serves no purpose.